roachtest: kv/gracefuldraining failed

roachtest.kv/gracefuldraining [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15670966?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15670966?buildTab=artifacts#/kv/gracefuldraining) on master @ [3aa5bdf40527e2f60b179094403b4302a1c2cbe1](https://github.com/cockroachdb/cockroach/commits/3aa5bdf40527e2f60b179094403b4302a1c2cbe1):


```
(assertions.go:363).Fail: 
	Error Trace:	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/kv.go:590
	            				main/pkg/cmd/roachtest/monitor.go:120
	            				golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78
	            				src/runtime/asm_amd64.s:1695
	Error:      	Received unexpected error:
	            	QPS of 5.00 at time 2024-06-15 08:46:06.251665142 +0000 UTC m=+10554.934669970 is below minimum allowable QPS of 1800.00
	            	(1) attached stack trace
	            	  -- stack trace:
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerKVGracefulDraining.func1.2
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/kv.go:573
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerKVGracefulDraining.func1.4
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/kv.go:590
	            	  | main.(*monitorImpl).Go.func1
	            	  | 	main/pkg/cmd/roachtest/monitor.go:120
	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	            	  | 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78
	            	  | runtime.goexit
	            	  | 	src/runtime/asm_amd64.s:1695
	            	Wraps: (2) QPS of 5.00 at time 2024-06-15 08:46:06.251665142 +0000 UTC m=+10554.934669970 is below minimum allowable QPS of 1800.00
	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
	Test:       	kv/gracefuldraining
(require.go:1357).NoError: FailNow called
test artifacts and logs in: /artifacts/kv/gracefuldraining/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-15670966/kv-gracefuldraining/1718441130292/1718441330705)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*kv/gracefuldraining.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39576

The workload exits with an error `result is ambiguous` early:

```
Error: ERROR: result is ambiguous: error=ba: Put [/Table/106/1/-8592365221218082830/0], EndTxn(parallel commit) [/Table/106/1/-8592365221218082830/0], [txn: 6d018e98], [can-forward-ts] RPC error: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable] [exhausted] (last error: ba: Put [/Table/106/1/-8592365221218082830/0], EndTxn(parallel commit) [/Table/106/1/-8592365221218082830/0], [txn: 6d018e98], [can-forward-ts] RPC error: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]) (SQLSTATE 40003)
```

This causes QPS to drop to zero shortly after the node is stopped, as there QPS inbound is now zero.

```
2024/06/15 08:46:04 cluster.go:692: test status: stopping nodes :6
teamcity-15670966-1718430614-34-n7cpu4: stopping
2024/06/15 08:46:05 kv.go:565: test worker status: workload command completed
```

The ambiguous error seems expected, or at least always possible with nodes being stopped? Presumably, the lease transferred away to another node but the EndTxn was issued just prior, the node shuts down and the error is returned to the client?

--- a/pkg/cmd/roachtest/tests/kv.go
+++ b/pkg/cmd/roachtest/tests/kv.go
@@ -540,6 +540,11 @@ func registerKVGracefulDraining(r registry.Registry) {
 			err := WaitFor3XReplication(ctx, t, t.L(), dbs[0])
 			require.NoError(t, err)
 
+			// TODO(baptist): Remove setting once #129427 is addressed.
+			if _, err := dbs[0].ExecContext(ctx, "SET CLUSTER SETTING kv.store_gossip.max_frequency = '0s'"); err != nil {
+				t.Fatalf("failed to disable load based splitting: %v", err)
+			}
+
 			t.Status("initializing workload")
 
 			// Initialize the database with a lot of ranges so that there are

--- a/pkg/cmd/roachtest/tests/kv.go
+++ b/pkg/cmd/roachtest/tests/kv.go
@@ -616,37 +616,11 @@ func registerKVGracefulDraining(r registry.Registry) {
 					t.Status("letting workload run with all nodes")
 					select {
 					case <-ctx.Done():
-						return
+						t.Fatalf("context cancelled while waiting")
 					case <-time.After(2 * time.Minute):
 					}
 				}
-				// Graceful drain and allow it to complete. The liveness record is
-				// updated at the beginning of the drain process, so by time the drain
-				// completes in ~5s all other nodes should "know" it is draining.
-				cmd := fmt.Sprintf("./cockroach node drain --certs-dir=%s --port={pgport%s} --self", install.CockroachNodeCertsDir, restartNode)
-				c.Run(ctx, option.WithNodes(restartNode), cmd)
-				// Simulate a hard network drop to this node prior to shutting it down.
-				// This is what we see in some customer environments. As an example, a
-				// docker container shutdown will also disappear from the network and
-				// drop all packets in both directions.
-				// TODO(baptist): Convert this to use a network partitioning
-				// utility function.
-				if !c.IsLocal() {
-					c.Run(ctx, option.WithNodes(restartNode), `sudo iptables -A INPUT -p tcp --dport 26257 -j DROP`)
-					c.Run(ctx, option.WithNodes(restartNode), `sudo iptables -A OUTPUT -p tcp --dport 26257 -j DROP`)
-				}
-				c.Stop(ctx, t.L(), option.DefaultStopOpts(), restartNode)
-				t.Status("letting workload run with one node down")
-				select {
-				case <-ctx.Done():
-					return
-				case <-time.After(1 * time.Minute):
-				}
-				// Clean up the iptables rule before restarting.
-				if !c.IsLocal() {
-					c.Run(ctx, option.WithNodes(restartNode), `sudo iptables -F`)
-				}
-				c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings(), restartNode)
+				drainWithIpTables(ctx, restartNode, c, t)
 				m.ResetDeaths()
 			}
 
@@ -666,6 +640,40 @@ func registerKVGracefulDraining(r registry.Registry) {
 			m.Wait()
 		},
 	})
+}
+
+// drainWithIpTables does a graceful drain and allows it to complete. The
+// liveness record is updated at the beginning of the drain process, so by time
+// the drain completes in ~5s all other nodes should "know" it is draining.
+func drainWithIpTables(
+	ctx context.Context, restartNode option.NodeListOption, c cluster.Cluster, t test.Test,
+) {
+	cmd := fmt.Sprintf("./cockroach node drain --certs-dir=%s --port={pgport%s} --self", install.CockroachNodeCertsDir, restartNode)
+	c.Run(ctx, option.WithNodes(restartNode), cmd)
+
+	// Simulate a hard network drop to this node prior to shutting it down. This
+	// is what we see in some customer environments. As an example, a docker
+	// container shutdown will also disappear from the network and drop all
+	// packets in both directions.
+	// TODO(baptist): Convert this to use a network partitioning utility.
+	if !c.IsLocal() {
+		c.Run(ctx, option.WithNodes(restartNode), `sudo iptables -A INPUT -p tcp --dport 26257 -j DROP`)
+		c.Run(ctx, option.WithNodes(restartNode), `sudo iptables -A OUTPUT -p tcp --dport 26257 -j DROP`)
+		// NB: We don't use the original context as it might be cancelled.
+		defer c.Run(context.Background(), option.WithNodes(restartNode), `sudo iptables -F`)
+	}
+	c.Stop(ctx, t.L(), option.DefaultStopOpts(), restartNode)
+
+	t.Status("letting workload run with one node down")
+	select {
+	case <-ctx.Done():
+		t.Fatalf("context cancelled while waiting")
+	case <-time.After(1 * time.Minute):
+	}
+
+	startOpts := option.DefaultStartOpts()
+	startOpts.RoachprodOpts.SkipInit = true
+	c.Start(ctx, t.L(), startOpts, install.MakeClusterSettings(), restartNode)
 }
 
 func registerKVSplits(r registry.Registry) {

--- a/pkg/cmd/roachtest/tests/kv.go
+++ b/pkg/cmd/roachtest/tests/kv.go
@@ -566,9 +566,10 @@ func registerKVGracefulDraining(r registry.Registry) {
 			// Three iterations, each iteration has a 3-minute duration.
 			desiredRunDuration := 10 * time.Minute
 			m.Go(func(ctx context.Context) error {
+				// Don't connect to the node we are going to shut down.
 				cmd := fmt.Sprintf(
 					"./cockroach workload run kv --duration=%s --read-percent=50 --follower-read-percent=50 --concurrency=200 --max-rate=%d {pgurl%s}",
-					desiredRunDuration, specifiedQPS, c.CRDBNodes())
+					desiredRunDuration, specifiedQPS, c.Range(1, nodes-1))
 				t.WorkerStatus(cmd)
 				defer func() {
 					t.WorkerStatus("workload command completed")

--- a/pkg/cmd/roachtest/tests/kv.go
+++ b/pkg/cmd/roachtest/tests/kv.go
@@ -534,6 +534,7 @@ func registerKVGracefulDraining(r registry.Registry) {
 			dbs := make([]*gosql.DB, nodes-1)
 			for i := range dbs {
 				dbs[i] = c.Conn(ctx, t.L(), i+1)
+				defer dbs[i].Close()
 			}
 
 			err := WaitFor3XReplication(ctx, t, t.L(), dbs[0])

